README
NAME

    Amazon::MWS - Perl interface to Amazon Marketplace Web Services

VERSION

    0.150

DESCRIPTION

    See Amazon::MWS::Client for the implementation of the low level Amazon
    MWS API.

    For the high level methods provided by the Amazon::MWS::Uploader you
    are going to need to supply database storage.

 Uploading products

    Basically you take the products which need to be uploaded to Amazon and
    create Amazon::MWS::XML::Product objects from them:

        use Amazon::MWS::XML::Product;
        use Amazon::MWS::Uploader;
        use Try::Tiny;
    
        my @upload;
    
        foreach my $product_object (@send_to_amazon) {
            my %prod = ....
    
            try {
                my $product = Amazon::MWS::XML::Product->new(%prod);
                push @upload, $product;
            } catch {
                warn "Failure with $prod{sku} : $_";
            };
         }

    Now you create an Amazon::MWS::XML::Uploader object and do the actual
    upload.

        my $uploader = Amazon::MWS::XML::Uploader->new(...);
    
        try {
            $uploader->products(\@upload);
            $uploader->upload;
        } catch {
            die "Uploader failure with: $_";
        };

    A product upload goes through the following steps:

    product

      Basic product data.

    inventory

      Product inventory count (required).

    price

      Product price (required).

    image

      Product images (required).

      Please note that only http:// links are allowed. If you pass https://
      links, they will be rejected by Amazon.

    variants

      Product variants (optional).

    You are not by any means finished with the upload, see "Resume
    uploads".

 Delete products

    Deleting products is much more straightforward, but you have still to
    care of tracking the upload process, see again "Resume uploads".

        my $uploader = Amazon::MWS::Uploader->new(...);
        $uploader->delete_skus(@skus);

 Resume uploads

    This is necessary to check whether the upload has been processed by
    Amazon.

        my $uploader = Amazon::MWS::XML::Uploader->new(...);
        $uploader->resume;

    The uploader also goes automatically through the steps for the product
    uploads.

MWS in practice

 Product price

    Every product uploaded needs a price of 0.01 or higher, otherwise you
    get the following error:

        0.00 price (standard or sales) will not be accepted.
        Please ensure that every SKU in your feed has a price at least equal to or greater than 0.01

 Shipping costs

    You need to configure the shipping costs in Amazon Seller Central, you
    can't pass them through MWS:

    https://sellercentral.amazon.com/gp/shipping/dispatch.html

 Stuck uploads

    There is no guarantee that Amazon finishes your uploads at all. We had
    uploads stuck for at least a week.

 Multiple marketplaces

    You can use this module and the uploader for multiple Amazon
    marketplaces. Please make sure that you disable Amazon's
    synchronisation between marketplaces.

    For marketplaces with a different currency you need to convert your
    price first.

    The list of marketplaces can be found at:

    http://docs.developer.amazonservices.com/en_US/dev_guide/DG_Endpoints.html

 Throttling and Quota

    With Amazon MWS you have to deal with Amazon throttling your uploads
    and imposing quotas.

    Possible reasons:

    Upload too often

    Stuck uploads

    Orders with orderlines

  Throttle Reponse

        <?xml version="1.0"?>
        <ErrorResponse xmlns="http://mws.amazonaws.com/doc/2009-01-01/">
          <Error>
          <Type></Type>
          <Code>RequestThrottled</Code>
          <Message>Request is throttled</Message>
        </Error>
        <RequestID>a7b39ee6-4f76-48ee-92f1-43bc54f693df</RequestID>
        </ErrorResponse>

  Quota Exceeded Error Response

        <?xml version="1.0"?>
        <ErrorResponse xmlns="http://mws.amazonaws.com/doc/2009-01-01/">
          <Error>
            <Type></Type>
            <Code>QuotaExceeded</Code>
            <Message>You exceeded your quota of 80.0 requests per 1 hour for operation Feeds/2009-01-01/GetFeedSubmissionList.  Your quota will reset on Thu Dec 18 07:39:00 UTC 2014</Message>
          </Error>
          <RequestID>5115e00d-35a8-4589-8083-f0ef998f76ef</RequestID>
        </ErrorResponse>

 Upload Errors

  Error 8572: Incorrect UPCs/EANs/...

       error: You are using UPCs, EANs, ISBNs, ASINs, or JAN codes
       that do not match the products you are trying to list. Using incorrect UPCs,
       EANs, ISBNs, ASINs or JAN codes is prohibited and it can result in your ASIN
       creation privileges being suspended or permanently removed.

    This happened in our case with products having self assigned EAN codes,
    but with a well known manufacturer. Apparently Amazon knows about the
    EAN ranges for this manufacturer.

  Error 18181

    The format of the tracking number passed to Amazon was invalid, e.g.
    wrong number of characters.

  Error 20013: Image file size

       error: Image file size: 13972730 bytes exceeds the the maximum allowed file size: 10485760 bytes.

    There is no point in using such big image files for Amazon.

  Error 99001: "brand_name"

    Brand is a required field for products.

Uploader Module

    Amazon::MWS::Uploader is an upload agent for Amazon::MWS.

XML Modules

    Generic Feed

      Amazon::MWS::XML::GenericFeed

    Feed

      Amazon::MWS::XML::Feed

    Product

      Amazon::MWS::XML::Product

    Address

      Amazon::MWS::XML::Address

    Order

      Amazon::MWS::XML::Order

    OrderlineItem

      Amazon::MWS::XML::OrderlineItem

ACKNOWLEDGEMENTS

    James Risner (additional API methods).

AUTHORS

    Paul Driver Phil Smith Marco Pessotto Stefan Hornburg (Racke)

 COPYRIGHT

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Amazon::MWS

0.800 NOT YET RELEASED

     [ENHANCEMENTS]

     * Set inventory to 0 if price is zero and upload.
       Upload image and variants feed if product is inactive (Marco Pessotto, GH #24).
     * Add customer_vat_id and is_business methods to Amazon::MWS::XML::Order (Marco Pessotto)
     * Exclude from get_products_with_mismatches the products with ok status (Marco Pessotto)
     * Treat HTTP response error code 401 as "Amazon::MWS::Exception::Response"
       (Stefan Hornburg/Racke)
     * Handle missing SKU in warning gracefully (Stefan Hornburg/Racke).
     * Add Travis CI configuration file (Stefan Hornburg/Racke).
     * Support for Finances API methods (Eric Ferguson).

0.150 Sat Sep 23 13:20:08 2017 CEST

     [BUG FIXES]

     * Fix error diagnostics for ack response (Marco Pessotto).
     * Don't skip failed variants for forced uploads (Stefan Hornburg/Racke).

     [ENHANCEMENTS]

     * Add API methods (James Risner):

       GetInboundGuidanceForASIN
       GetInboundGuidanceForSKU

0.141 Mon Dec  5 13:30:14 2016 CET

     [BUG FIXES]

     * Handle hash references for address values (Marco Pessotto).

     [DOCUMENTATION]

     * Add documentation on Amazon error 18181 (Stefan Hornburg/Racke).

0.140 Mon Nov  7 10:27:33 2016 CET

     [ENHANCEMENTS]

     * Add API methods provided by James Risner (GH #13, #14):

       GetLowestPricedOffersForSKU
       GetLowestPricedOffersForASIN
       GetMyPriceForSKU
       GetMyPriceForASIN
       GetMyFeesEstimate

     [BUG FIXES]

     * Fix signature of API method GetServiceStatus (Stefan Hornburg/Racke).

0.132 Sun Oct 23 09:01:46 2016 CEST

     [ENHANCEMENTS]

     * Add GetPrepInstructionsForASIN and GetPrepInstructionsForSKU
       API methods provided by James Risner.

0.131 Wed May 25 09:50:20 2016 CEST

     [ENHANCEMENTS]

     * Allow to limit the product upload steps (Marco Pessotto).
     * Removed confirmed => 1 filters to get the orders waiting for shipping
      (Marco Pessotto).
     * Make _get_pending_jobs method of the uploader public
       (Marco Pessotto).
     * Allow additional parameters for resume method and keep the jobs
	   in predictable order (Marco Pessotto).

     [DOCUMENTATION]

     * Add sections about uploading products, deleting products
       and resuming uploads to the main documentation (Stefan Hornburg/Racke).
     * Add note about image links to Product class documentation
       (Stefan Hornburg/Racke).
     * Add comment about usage of Amazon::MWS::Client to the documentation.
       (Stefan Hornburg/Racke).

0.130 Sat Apr  2 08:08:26 2016 CEST

     [ENHANCEMENTS]

     * Add method to retrieve skus with warnings (Marco Pessotto).
     * Add color and size mismatches, also handle errors in German
       (Marco Pessotto).

     [BUG FIXES]

     * Strip trailing slash from endpoint
       (Marco Pessotto, jordanmhiller, GH #9).

     [DOCUMENTATION]

     * Add possible explanation for error code 8572 (Stefan Hornburg/Racke).

0.120 Tue Mar 29 07:35:46 2016 CEST

     [ENHANCEMENTS]

     * Permit arguments to resume method in Uploader class
       (Marco Pessotto).
     * Do not perform checks on products if checked_products is set
       (Marco Pessotto).
     * Add get_products_with_mismatches method (Marco Pessotto).
     * Improve separation into first and last name (Marco Pessotto).

     [DOCUMENTATION]

     * Add possible explanation for error code 8572 (Stefan Hornburg/Racke).

0.116 Sat Dec 19 15:12:10 2015 CET

     [ENHANCEMENTS]

     * Return the shipping category level as shipping_method
       (Marco Pessotto).

0.115 Sat Nov  7 12:38:49 2015 CET

     [ENHANCEMENTS]

     * Add purge_old_jobs method (Marco Pessotto).

     [BUG FIXES]

     * Exclude MYMETA.* files from distribution
       (Alexandr Ciornii, Stefan Hornburg, GH #7).

0.114 Mon Oct  5 12:22:23 2015 CEST

     [BUG FIXES]

     * Prevent crash from empty ItemPrice or ShippingPrice caused by
       cancelled items (Marco Pessotto).

0.113 Wed Jul 22 09:31:32 2015 CEST

     [BUG FIXES]

     * Add prerequisite on namespace::clean (Stefan Hornburg, GH #6).

     [ENHANCEMENTS]

     * Add quiet accessor to Uploader (Marco Pessotto).

0.112 Mon Jul 20 17:29:42 2015 CEST

     * Initial release.


lib/Amazon/MWS.pm
package Amazon::MWS;

=head1 NAME

Amazon::MWS - Perl interface to Amazon Marketplace Web Services

=head1 VERSION

0.150

=cut

our $VERSION = '0.150';

=head1 DESCRIPTION

See L<Amazon::MWS::Client> for the implementation of the low level
Amazon MWS API.

For the high level methods provided by the L<Amazon::MWS::Uploader>
you are going to need to supply database storage.

=head2 Uploading products

Basically you take the products which need to be uploaded to Amazon
and create L<Amazon::MWS::XML::Product> objects from them:

    use Amazon::MWS::XML::Product;
    use Amazon::MWS::Uploader;
    use Try::Tiny;

    my @upload;

    foreach my $product_object (@send_to_amazon) {
        my %prod = ....

        try {
            my $product = Amazon::MWS::XML::Product->new(%prod);
            push @upload, $product;
        } catch {
            warn "Failure with $prod{sku} : $_";
        };
     }

Now you create an L<Amazon::MWS::XML::Uploader> object and do the
actual upload.

    my $uploader = Amazon::MWS::XML::Uploader->new(...);

    try {
        $uploader->products(\@upload);
        $uploader->upload;
    } catch {
        die "Uploader failure with: $_";
    };

A product upload goes through the following steps:

=over 4

=item product

Basic product data.

=item inventory

Product inventory count (required).

=item price

Product price (required).

=item image

Product images (required).

Please note that B<only http:// links> are allowed. If you pass https://
links, they will be rejected by Amazon.

=item variants

Product variants (optional).

=back

You are not by any means finished with the upload, see L</Resume uploads>.

=head2 Delete products

Deleting products is much more straightforward, but you have still to
care of tracking the upload process, see again L</Resume uploads>.

    my $uploader = Amazon::MWS::Uploader->new(...);
    $uploader->delete_skus(@skus);

=head2 Resume uploads

This is necessary to check whether the upload has been
processed by Amazon.

    my $uploader = Amazon::MWS::XML::Uploader->new(...);
    $uploader->resume;

The uploader also goes automatically through the steps for the
product uploads.

=head1 MWS in practice

=head2 Product price

Every product uploaded needs a price of 0.01 or higher, otherwise you
get the following error:

    0.00 price (standard or sales) will not be accepted.
    Please ensure that every SKU in your feed has a price at least equal to or greater than 0.01

=head2 Shipping costs

You need to configure the shipping costs in Amazon Seller Central, you can't pass them
through MWS:

L<https://sellercentral.amazon.com/gp/shipping/dispatch.html>

=head2 Stuck uploads

There is no guarantee that Amazon finishes your uploads at all. We had uploads
stuck for at least a week.

=head2 Multiple marketplaces

You can use this module and the uploader for multiple Amazon marketplaces.
Please make sure that you disable Amazon's synchronisation between marketplaces.

For marketplaces with a different currency you need to convert your price first.

The list of marketplaces can be found at:

L<https://docs.developer.amazonservices.com/en_US/dev_guide/DG_Endpoints.html>

=head2 Throttling and Quota

With Amazon MWS you have to deal with Amazon throttling your uploads and
imposing quotas.

Possible reasons:

=over 4

=item Upload too often

=item Stuck uploads

=item Orders with orderlines

=back

=head3 Throttle Reponse

    <?xml version="1.0"?>
    <ErrorResponse xmlns="http://mws.amazonaws.com/doc/2009-01-01/">
      <Error>
      <Type></Type>
      <Code>RequestThrottled</Code>
      <Message>Request is throttled</Message>
    </Error>
    <RequestID>a7b39ee6-4f76-48ee-92f1-43bc54f693df</RequestID>
    </ErrorResponse>

=head3 Quota Exceeded Error Response

    <?xml version="1.0"?>
    <ErrorResponse xmlns="http://mws.amazonaws.com/doc/2009-01-01/">
      <Error>
        <Type></Type>
        <Code>QuotaExceeded</Code>
        <Message>You exceeded your quota of 80.0 requests per 1 hour for operation Feeds/2009-01-01/GetFeedSubmissionList.  Your quota will reset on Thu Dec 18 07:39:00 UTC 2014</Message>
      </Error>
      <RequestID>5115e00d-35a8-4589-8083-f0ef998f76ef</RequestID>
    </ErrorResponse>

=head2 Upload Errors

=head3 Error 8572: Incorrect UPCs/EANs/...

   error: You are using UPCs, EANs, ISBNs, ASINs, or JAN codes
   that do not match the products you are trying to list. Using incorrect UPCs,
   EANs, ISBNs, ASINs or JAN codes is prohibited and it can result in your ASIN
   creation privileges being suspended or permanently removed.

This happened in our case with products having self assigned EAN codes, but
with a well known manufacturer. Apparently Amazon knows about the EAN ranges
for this manufacturer.

=head3 Error 18181

The format of the tracking number passed to Amazon was invalid, e.g.
wrong number of characters.

=head3 Error 20013: Image file size

   error: Image file size: 13972730 bytes exceeds the the maximum allowed file size: 10485760 bytes.

There is no point in using such big image files for Amazon.

=head3 Error 99001: "brand_name"

Brand is a required field for products.

=head2 Temporary errors

The following error codes are issued for internal errors: 5002, 14001, 20003, 159144.
There is nothing you can do about, so ignorance is the best way to deal with them.

=head3 Error 20003

   error: Submitting image updates for SKU 0815 is temporarily unavailable. Please retry your submission at a later date.

=head2 Other problems

It can take a while for additional images to show up on the Amazon website after a successful
upload.

=head1 Uploader Module

L<Amazon::MWS::Uploader> is an upload agent for Amazon::MWS.

=head1 XML Modules

=over 4

=item Generic Feed

L<Amazon::MWS::XML::GenericFeed>

=item Feed

L<Amazon::MWS::XML::Feed>

=item Product

L<Amazon::MWS::XML::Product>

=item Address

L<Amazon::MWS::XML::Address>

=item Order

L<Amazon::MWS::XML::Order>

=item OrderlineItem

L<Amazon::MWS::XML::OrderlineItem>

=back

=head1 ACKNOWLEDGEMENTS

James Risner (additional API methods).
Eric Ferguson (support for Finances API).

=head1 AUTHORS

Paul Driver
Phil Smith
Marco Pessotto
Stefan Hornburg (Racke)

=head2 COPYRIGHT

This is free software; you can redistribute it and/or modify it under the same terms
as the Perl 5 programming language system itself.

=cut

1;
